### DISCUSSION TITLE
What would be needed for easy adoption for the project or parts of it?

### DISCUSSION THREAD

**Discussion by johonkanen** [johonkanen/ac_inout_psu#31] [General]
In order to make the code more usable, you can either add an issue or discuss possible problems. There are bound to be some since this project was started from empty file quite recently.

**Comment by lukipedio** [comment]
Even if there is a lot of information on the project companion website, the project goal is not completely clear to me, since I am not aware of all the effort behind the design of a grid-connected power supply.
As far as I see this all started as a thesis on the subject and it is evolving further,
Since the project is partially open-sourced (hardware design is not), I think it will stop its adoption, if this is the goal.
The firmware design is interesting for me for two reasons mainly:
- code structure: the use of ubiquitous use of records is quite intricating. An attempt to use objects in a non-object-oriented language.
- single IPs: UDP over ethernet, filters, UART are all pieces of code of some interest. I don't know if some of the digital processing and filtering is new or any way of some interest for the community of power supply designers.

Unit testing approach is another point of interest. I praise your attempt to implement it from the ground up.
I think VUnit has already faced the issues you are planning to cope with. 
So I propose to keep active the https://github.com/lukipedio/ac_inout_psu/tree/feature/vunit branch so that it can be integrated at some point in the main project if it demonstrates itself as a valid approach.

**Reply by johonkanen** [reply]
I also think that the IPs and the code structure are what should be made to be as easily usable as possible. The uart for example allows access to uart with less than 20 lines of source code with both receive and transmit of data being accessible with single procedure calls after the module is initialized. The ram reader allows very simple and intuitive ram interface control and sharing the ram control port with several modules with very little effort. The math library gives simple access to a multiplication, filtering and dynamic simulation with just a few lines of code. The project as a whole is probably more of a novelty value as not many will be building a bidirectional power supplies even fewer with full FPGA control.

The power electronics hardware has not been built yet, but I will put up the schematics probably with the next blog post as I am doing the modeling and FPGA simulation of it. I intend to release gerbers and bill of materials as soon as electronics have been built.

**Reply by johonkanen** [reply]
The project is currently built on both the power supply control board as well as the cyclone 10lp evaluation kit to make it accessible without need to order custom pcb:s

**Comment by lukipedio** [comment]
I am an electronics engineer, but I have never worked with power supply design (even if my company has also a HV power supply division for high-energy physics). A short introduction of applications and a sketch diagram of connections could be useful to non experts.
I miss the bidirectional nature of the supply: I think the application is to generate AC voltage from a DC  input (power cells or whatever generates a DC voltage). In a grid connected power supply, external provided power supply () can also be used to charge a local battery system?

**Reply by johonkanen** [reply]
The bidirectionality allows it to be used both as a controllable load as well as a voltage supply for either ac or dc loads. The whole power supply has three switch mode converters, a grid side inverter for interfacing the AC grid to high voltage DC, output inverter for DC to AC conversion and a transformer isolated DC/DC converter to galvanically isolate the two inverters from each other. The whole system operation and its modeling and FPGA simulation will be next topic so which I will hopefully have up next week.
